File: frontend/modules/trigger/infraestructure/strapi/trigger-model.ts
import { FileData, fileUrlToFileData } from '../../../shared/infraestructure/strapi/file'
import { Trigger, TriggerCompare } from '../../domain/trigger'

export interface TriggerAttributes {
  title: string
  uuid: string
  input: string
  compare: string
  output_text?: string
  output_audio?: FileData|null
  discord_server_id: string
}

export interface TriggerModel {
  id?: number
  attributes: TriggerAttributes
}

export function triggerAggregateToModel(trigger: Trigger): TriggerModel {
  return {
    attributes: {
      title: trigger.title,
      uuid: trigger.uuid,
      compare: trigger.compare,
      input: trigger.input,
      output_audio: fileUrlToFileData(trigger.outputAudio),
      output_text: trigger.outputText,
      discord_server_id: trigger.discordServerId
    }
  }
}

export function triggerModelToAggregate(model: TriggerModel): Trigger {
  return Trigger.fromObject({
    title: model.attributes.title,
    uuid: model.attributes.uuid,
    compare: model.attributes.compare as TriggerCompare,
    discordServerId: model.attributes.discord_server_id,
    input: model.attributes.input,
    outputAudio: model.attributes.output_audio?.data?.attributes?.url,
    outputText: model.attributes.output_text
  })
}

File: context/src/main/kotlin/com/usadapekora/context/infraestructure/strapi/trigger/TriggerModel.kt
package com.usadapekora.context.infraestructure.strapi.trigger

import com.usadapekora.context.infraestructure.strapi.resources.Model
import kotlinx.serialization.Serializable
import com.usadapekora.context.domain.trigger.Trigger

@Serializable
data class TriggerModel (
    val id: Int,
    val attributes: TriggerAttributes
): Model() {
    companion object {
        fun fromAggregate(trigger: Trigger): TriggerModel = TriggerModel(
            id = trigger.id.value.toInt(),
            attributes = TriggerAttributes(
                input = trigger.input.value,
                compare = trigger.compare.value,
                discord_server_id = trigger.discordGuildId.value,
                output_text = trigger.outputText.value,
                output_audio = null // TODO: resolve output audio from aggregate to model
            )
        )
    }

    fun toAggregate(): Trigger = Trigger.fromPrimitives(
        id = id.toString(),
        title = "no title",
        input = attributes.input,
        compare = attributes.compare,
        outputText = attributes.output_text,
        discordGuildId = attributes.discord_server_id
    )
}
